core/module_specifier.rs

```diff
@@ -76,33 +79,78 @@ impl ModuleSpecifier {
       // problem lies somewhere else.
       Err(err) => Err(InvalidUrl(err))?,
     };
 
     Ok(ModuleSpecifier(url))
   }
 
-  /// Takes a string representing a path or URL to a module, but of the type
-  /// passed through the command-line interface for the main module. This is
-  /// slightly different than specifiers used in import statements: "foo.js" for
-  /// example is allowed here, whereas in import statements a leading "./" is
-  /// required ("./foo.js"). This function is aware of the current working
-  /// directory and returns an absolute URL.
-  pub fn resolve_root(
-    root_specifier: &str,
+  /// Converts a string representing an absulute URL into a ModuleSpecifier.
+  pub fn resolve_url(
+    url_str: &str,
   ) -> Result<ModuleSpecifier, ModuleResolutionError> {
-    let url = match Url::parse(root_specifier) {
-      Ok(url) => url,
-      Err(..) => {
-        let cwd = std::env::current_dir().unwrap();
-        let base = Url::from_directory_path(cwd).unwrap();
-        base.join(&root_specifier).map_err(InvalidUrl)?
-      }
-    };
+    Url::parse(url_str)
+      .map(ModuleSpecifier)
+      .map_err(ModuleResolutionError::InvalidUrl)
+  }
 
-    Ok(ModuleSpecifier(url))
+  /// Takes a string representing either an absolute URL or a file path,
+  /// as it may be passed to deno as a command line argument.
+  /// The string is interpreted as a URL if it starts with a valid URI scheme,
+  /// e.g. 'http:' or 'file:' or 'git+ssh:'. If not, it's interpreted as a
+  /// file path; if it is a relative path it's resolved relative to the current
+  /// working directory.
+  pub fn resolve_url_or_path(
+    specifier: &str,
+  ) -> Result<ModuleSpecifier, ModuleResolutionError> {
+    if Self::specifier_has_uri_scheme(specifier) {
+      Self::resolve_url(specifier)
+    } else {
+      Self::resolve_path(specifier)
+    }
+  }
+
+  /// Converts a string representing a relative or absolute path into a
+  /// ModuleSpecifier. A relative path is considered relative to the current
+  /// working directory.
+  fn resolve_path(
+    path_str: &str,
+  ) -> Result<ModuleSpecifier, ModuleResolutionError> {
+    let path = current_dir().unwrap().join(path_str);
+    Url::from_file_path(path)
+      .map(ModuleSpecifier)
+      .map_err(|()| ModuleResolutionError::InvalidPath)
+  }
+
+  /// Returns true if the input string starts with a sequence of characters
+  /// that could be a valid URI scheme, like 'https:', 'git+ssh:' or 'data:'.
+  ///
+  /// According to RFC 3986 (https://tools.ietf.org/html/rfc3986#section-3.1),
+  /// a valid scheme has the following format:
+  ///   scheme = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+  ///
+  /// We additionally require the scheme to be at least 2 characters long,
+  /// because otherwise a windows path like c:/foo would be treated as a URL,
+  /// while no schemes with a one-letter name actually exist.
+  fn specifier_has_uri_scheme(specifier: &str) -> bool {
```

Comment:
I was gonna argue that trying path first before parsing URL was to handle Windows path prefixed with disk, but as always your solution is much more elegant :+1: kudos @piscisaureus